components/server/src/auth/bearer-authenticator.ts

```diff
@@ -79,19 +80,51 @@ export class BearerAuth {
         if (!token) {
             throw createBearerAuthError("missing Bearer token");
         }
-        const userAndToken = await this.userDB.findUserByGitpodToken(token, GitpodTokenType.API_AUTH_TOKEN);
-        if (!userAndToken) {
-            throw createBearerAuthError("invalid Bearer token");
-        }
 
-        // hack: load the user again to get ahold of all identities
-        // TODO(cw): instead of re-loading the user, we should properly join the identities in findUserByGitpodToken
-        const user = (await this.userDB.findUserById(userAndToken.user.id))!;
+        let user: User;
+        let scopes: string[];
+        // We handle two types of Bearer tokens:
+        //  1. Personal Access Tokens which are prefixed with `gitpod_pat_`
+        //  2. Old(er) access tokens which do not have any specific prefix.
+        if (token.startsWith("gitpod_pat_")) {
+            try {
+                const hash = await this.validatePersonalAccessToken(token);
+
+                const pat = await this.personalAccessTokenDB.getByHash(hash);
+
+                if (!pat) {
+                    // No PAT exists for this hash. We must not leak that information so return a generic error.
+                    throw createBearerAuthError("Invalid personal access token");
+                }
+
+                const userByID = await this.userDB.findUserById(pat.userId);
+                if (!userByID) {
+                    // No corresponding user for the PAT. We must not leak that information so return a generic error.
```

Comment:
The comment should actually be moved to catch clause. Here, we could throw any error, actually.
Noting this because the current state feels like a misunderstanding/potential bug.

Reply:
How would this be handled in the catch? `findUserById` returns undefined, or User which wouldn't throw, no?

Reply:
Sorry for the confusion: Throwing here is fine, it's more that it feels like double-addressing the same concern: Once inside the try-tach (where I think it belongs), and once at the throw side.

For instance, one could imagine throwing precise errors on the throw-side (maybe with a hint that those are expected to be caught below), in combination with a catch that logs the unterlying error + sending a new, generic error up the call chain.

Reply:
The reason for the try-catch wrap is I want to ensure that even when runtime exceptions occur, they never make it to the end-user. Yes, we could throw for the specific cases but we can also get runtime exceptions from the connection to DB failing etc.

I'm a bit confused as to what you want me to change here. Would you have an example?

Reply:
Sth ala
```
try {
   throw new Error("Cannot find user with ID");
} ctach (err) {
  log.error({ userId }, "some error with PATs", err);
  throw createBearerAuthError("Invalid personal access token");
}
```
So keep specific errors for debugging, but do not forward them by keeping the generic catch-all, for the reasons you outlined.
But leave it up to you to decide: approved the PR! :pray:

Reply:
Adjusted, hopefully it matches what you expected. If not, happy to follow-up on this.

Reply:
Nice!